쥬스 메이커 [STEP 1] Toy, uemu

ue-mu opened the pull request:
@juun97
안녕하세요 릴라, STEP 1 제출합니다.
잘 부탁드려요!!

## 🤔고민했던점
- 음료에 따른 분기처리 방법과 storage에 과일의 갯수를 소모시키는 방법을 고민했습니다.
  - 먼저 어떤 음료인지 분기처리를 하는 manufactureFruitJuice 함수를 만들고 음료에 따라 어떤 과일을 소모하는지 확인할 수 있는 branchHandler함수를 만들어 각 과일의 갯수를 담고있는 변수에 접근하여 갯수를 변경하도록 해결했습니다.
  
<br></br>
 ## 🤩조언을 얻고싶은 부분
- 네이밍과 컨벤션이 적절한지 궁금합니다.
- 함수의 기능 분리가 적절한지 궁금합니다.
- 오류 처리를 따로 구현하지 않고 조건문으로 과일 개수가 부족한 경우를 나눴는데 적절한 방법이였는지 궁금합니다.
- enum타입을 지정할때 FruitStorage과 FruitJuice를 파일을 각자 생성해서 관리하는 것과 한 곳에서 관리하는 것 중 어떤 방법이 더 적절한지 궁금합니다.

ue-mu committed:
feat: 과일 및 과일쥬스 열거형 생성

Developer-Groo committed:
feat: FruitStore 과일 변수 생성 및 과일 재고 소모하는 reduceFruitQuantity함수 생성

ue-mu committed:
feat: FruitStore 수량 체크하는 함수 생성

Developer-Groo committed:
feat: JuiceMaker구조체에서 입력받을 branchHandler함수 생성

ue-mu committed:
feat: brabchHandler 함수의 매개변수 타입 변경 및 과일 재고 변경 함수 생성

Developer-Groo committed:
feat: JuiceMaker구조체에서 쥬스생성하는 함수 구현

ue-mu committed:
feat: JuiceMaker 구조체 내의 함수명 수정 및 과일 수량 변경 함수 생성

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
-
 // 과일 저장소 타입
 class FruitStore {
+    static let singleTon = FruitStore()
```
1. struct 와 class 중에서 class 를 사용하신 이유가 있을까요?
2. 싱글톤 객체의 인스턴스의 네이밍은 보통 `shared` 를 사용합니다.

ue-mu replied:
싱글톤 객체의 인스턴스의 네이밍을 `shared` 로 수정 하였습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
-import Foundation
-
 // 과일 저장소 타입
 class FruitStore {
```
1. struct 와 class 중에서 class 를 사용하신 이유가 있을까요?
2. 상속이 필요한 객체일까요? 필요하지 않다면 어떻게 처리해줄 수 있을까요

ue-mu replied:
1. 과일 저장소 인스턴스를 생성했을 때 과일 재고가 어디서 접근하든 동일해야 하기 때문에 싱글톤 패턴을 사용하기 위해서 class를 사용했습니다.
2. 상속이 필요없는 객체라고 생각돼서 final 키워드를 사용하여 처리했습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
+    private (set) var banana = 10
+    private (set) var pineapple = 10
+    private (set) var kiwi = 10
+    private (set) var mango = 10
```
private set 을 잘 활용해주셨네요.
다른 접근 제어 키워드는 무엇 무엇이 있고 각각의 키워드는 어떤 접근 제어 레벨을 가지고 있나요?

또한 swift 의 기본 접근제어자는 무엇일까요?

ue-mu replied:
가장 낮은 접근 레벨로 private은 정의한 블록 내부에서만 접근 가능하고
다음 레벨에는 fileprivate로 하나의 파일 내에서만 접근 가능하며
그 다음 에는 기본 접근제어자인 internal이 있고 internal레벨로 선언되면 해당 모듈 전체에서 사용 가능합니다.
가장 높은 레벨의 접근 제어자는 open 과 public으로 모듈 외부에서도 접근이 가능합니다.
open과 public의 차이 점으로는 open은 다른 모듈에서 오버라이드와 서브클래싱이 가능하지만 public 접근자로 선언된 것은 다른 모듈에서는 오버라이드와 서브클래싱이 불가능 합니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStorage.swift:
```diff
@@ -0,0 +1,13 @@
+//  Copyright © yagom academy. All rights reserved.
+//
+
+enum FruitStorage {
+    case strawberry
+    case banana
+    case pineapple
+    case kiwi
+    case mango
+}
```
현재 타입의 네이밍으로는 Fruit 의 저장소를 의미하는 것 같습니다만, 구현해주신 사항으로는 Fruit 의 종류만을 가지고 있네요.
좀 더 적절한 네이밍이 필요할 것 같습니다

ue-mu replied:
네이밍 수정했습니다!

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitJuice.swift:
```diff
@@ -0,0 +1,18 @@
+//  Copyright © yagom academy. All rights reserved.
+//
+
+enum FruitJuice {
+    case strawberryJuice
+    case bananaJuice
+    case kiwiJuice
+    case pineappleJuice
+    case mangoJuice
+}
+
+enum MixFruitJuice {
+    case strawberryBananaJuice
+    case mangoKiwiJuice
+}
```
두 타입 모두 FruitJuice 의 종류인데 분리해서 만드신 이유가 있을까요?

ue-mu replied:
초기에는 분리해서 단일음료와 섞은음료의 차감방식을 나눠주려고 했는데
굳이 그럴필요가 없을거 같아서 하나의 FruitJuice로 합쳤습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
+        }
+    }
+    
+    private func checkQuantity(fruit: Int, check: FruitStorage, number: Int) {
```
fruit 의 갯수는 FruitStore 에서 관리하고 있는데 외부에서 주입받아야 하는 이유가 있을까요?
FruitStore 가 아닌 다른 타입에서 과일의 재고를 직접적으로 알아야 할 필요가 있나요?

ue-mu replied:
fruit의 갯수를 외부에서 주입받을 필요가 없다고 생각되어서
checkQuantity 함수의 로직을 수정해보았습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
+        }
+    }
+    
+    func branchHandler(fruit: FruitStorage, number: Int) {
```
네이밍이 모호한 것 같습니다. 좀 더 직관적인 네이밍이 필요할 것 같네요.

ue-mu replied:
코드를 수정하는 과정에서 branchHandler가 불필요하다고 느껴 삭제 하였습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
+    
+    private func reduceFruitQuantity(in fruit: FruitStorage, with number: Int) {
+        switch fruit {
+        case .strawberry:
+            strawberry -= number
+            print("딸기 재고: \(strawberry)개")
+        case .banana:
+            banana -= number
+            print("바나나 재고: \(banana)개")
+        case .pineapple:
+            pineapple -= number
+            print("파인애플 재고: \(pineapple)개")
+        case .kiwi:
+            kiwi -= number
+            print("키위 재고: \(kiwi)개")
+        case .mango:
+            mango -= number
+            print("망고 재고: \(mango)개")
+        }
```
동일한 로직들이 반복되고 있는데요, 어떻게 하면 한 번에 관리해줄 수 있을까요?

ue-mu replied:
동일한 로직을 관리하기 위해서 전체 코드의 구조를 수정하였습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,80 @@
+        }
+    }
     
+    func changeQuantity(fruit: FruitStorage, number: Int) {
+        switch fruit {
+        case .strawberry:
+            strawberry = number
+            print("딸기의 재고가 \(number)개로 변경되었습니다.")
+        case .banana:
+            banana = number
+            print("바나나의 재고가 \(number)개로 변경되었습니다.")
+        case .pineapple:
+            pineapple = number
+            print("파인애플의 재고가 \(number)개로 변경되었습니다.")
+        case .kiwi:
+            kiwi = number
+            print("키위의 재고가 \(number)개로 변경되었습니다.")
+        case .mango:
+            mango = number
+            print("망고의 재고가 \(number)개로 변경되었습니다.")
+        }
```
branchHandler 에서 과일에 따른 분기처리를 해주고 있는데, 이 메서드에서 분기처리를 또 해주는 이유가 있을까요?

ue-mu replied:
코드를 전체적으로 수정하는 과정에서 changeQuantity가 필요하지 않다고 생각되어 코드를 수정 하였습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,40 @@
 struct JuiceMaker {
+    private let fruitStore = FruitStore.singleTon
+    
+     func manufactureFruitJuice(in fruitJuice: FruitJuice) {
+        switch fruitJuice {
+        case .strawberryJuice:
+            fruitStore.branchHandler(fruit: .strawberry, number: 16)
+        case .bananaJuice:
+            fruitStore.branchHandler(fruit: .banana, number: 2)
+        case .kiwiJuice:
+            fruitStore.branchHandler(fruit: .kiwi, number: 3)
+        case .pineappleJuice:
+            fruitStore.branchHandler(fruit: .pineapple, number: 2)
+        case .mangoJuice:
+            fruitStore.branchHandler(fruit: .mango, number: 3)
+        }
+    }
+    
+    func manufactureMixFruitJuice(in mixFruitJuice: MixFruitJuice) {
+        switch mixFruitJuice {
+        case .strawberryBananaJuice:
+            fruitStore.branchHandler(fruit: .strawberry, number: 10)
+            fruitStore.branchHandler(fruit: .banana, number: 1)
+        case .mangoKiwiJuice:
+            fruitStore.branchHandler(fruit: .mango, number: 2)
+            fruitStore.branchHandler(fruit: .kiwi, number: 1)
+        }
```
위에서도 코멘트 드렸지만 현재 동일한 코드들이 계속해서 반복되고 있습니다.
물론 case 가 몇개 되지 않는경우에는 큰 문제가 되지 않습니다. 하지만 과일이나 주스의 종류가 늘어난다면 그 때마다 관련 코드들을 모두 수정을 해줄 필요가 있겠네요.

case 가 늘어나더라도 로직은 변경하지 않고 사용할 수 있는 방법을 찾아볼까요?

ue-mu replied:
이 피드백 또한 위와 동일한 이유로 코드를 새롭게 수정하였습니다.

juun97 reviewed (commented):
Toy, uemu STEP 1 고생하셨습니다 😃
빠르게 PR 보내주신 것에 비해 잘 수행 해주셨네요 👍🏻
좀 더 개선의 여지가 필요한 부분은 코멘트로 남겨드렸습니다.

## 🤩조언을 얻고싶은 부분
> 네이밍과 컨벤션이 적절한지 궁금합니다.

대체적으로 잘 작성 해주셨습니다만 몇몇 네이밍이 모호한 부분이 있는 것 같아 코멘트로 남겨드렸습니다.

> 함수의 기능 분리가 적절한지 궁금합니다.

하나의 함수가 하나의 기능만을 하도록 분리를 하려 한 흔적이 보였고, 실제로 잘 분리가 되있는 것 같네요 👍🏻

> 오류 처리를 따로 구현하지 않고 조건문으로 과일 개수가 부족한 경우를 나눴는데 적절한 방법이였는지 궁금합니다.

재고를 기준으로 오류처리를 해주시고 계십니다.
다만 `throw` 나 `Result 타입`을 통해 명시적으로 `Error` 를 보여준다면 코드의 직관성이 더 높아질 것 같습니다.


> enum타입을 지정할때 FruitStorage과 FruitJuice를 파일을 각자 생성해서 관리하는 것과 한 곳에서 관리하는 것 중 어떤 방법이 더 적절한지 궁금합니다.

상황에 따라 다르긴 한데요, 보통 각각의 파일에서 관리를 해주는 것이 좋습니다. 한개의 파일에 모두 관리를 하게 된다면 지금같이 소규모 프로젝트에선 큰 문제가 되지 않지만 프로젝트의 규모가 커지게 된다면 해당 파일이 어디에 위치하는지 파악하기가 힘들게 됩니다. 

비슷한 기능이라 관리를 해주고 싶다면 폴더그룹을 만들어 관리를 해주는 방법이 있을겁니다.

한 개의 파일에서 같이 관리를 하는 예시를 드리겠습니다.

A, B 타입이 존재한다고 가정할 때 A 타입이 B 타입에서만 사용이 되고 외부에서는 사용될일이 없는 경우 한 개의 파일에서 관리를 하기도 합니다.

Developer-Groo committed:
feat: 열거형 타입 수정

-FruitStorage에서 Fruit으로 타입명 변경
-Fruit타입에 String연관값 추가
-FruitJuice타입에 juiceRecipe계산속성 추가

ue-mu committed:
feat: FruitStore내의 각 과일의 프로퍼티를 dictonary로 수정 및 changeFruitQuantity함수 수정

Developer-Groo committed:
feat: 에러타입선언 및 재고확인 함수 수정

-JuiceMakerError타입 선언
-재고를 확인하고 부족할시 에러를 발생시키는 compareFruitQuantity함수 구현

Developer-Groo committed:
refactor: 재고확인함수 네이밍 수정

ue-mu committed:
feat: reduceFruitQuantity 함수 수정 및 brancHandler 함수 삭제

Developer-Groo committed:
feat: JuiceMaker구조체의 manufactureFruitJuice함수 수정

ue-mu committed:
feat: FruitStore 클래스에 final 키워드 추가

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -11,23 +11,14 @@ class FruitStore {
+                fruitQuantity[$0] = try checkQuantity(number: number, quantity: $1)
+            } catch {
+                throw JuiceMakerError.outOfStock
+            }
```
클로저의 단축은 개발자의 입장에서는 간편할 수 있습니다만 무엇을 의미하는지 파악하기 힘들 수도 있습니다.
$0, $1 같은 표현 보다는 파라미터의 네이밍을 직접적으로 명시 해주세요

ue-mu replied:
네이밍을 직접적으로 명시해 보았습니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,35 +6,13 @@
+             try fruitStore.reduceFruitQuantity(in: fruitJuice)
+         } catch {
+             print("과일의 재고가 부족하여 과일쥬스를 제조할수 없습니다.")
+         }
```
catch한 error 를 활용을 하고 있지 않네요.

1. LocalizedError 에 대해 알아볼까요?
2. try 를 do catch 이외에도 어떻게 사용할 수 있는지 알아볼까요?

ue-mu replied:
1. LocalizedError는 오류 및 오류가 발생한 이유를 설명하는 메시지를 제공하는 특수 오류입니다.
2. try? 는 리턴값을 옵셔널 타입 혹은 nil로 받을 수 있게 하고 try!는에러 발생시 crash 가 나고 반환타입은 non-optional 입니다.

juun97 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -31,7 +31,7 @@ class FruitStore {
+    private func checkQuantity(number: Int, quantity: Int) throws -> Int {
         if number - quantity >= 0 {
             return number - quantity
         } else {
```
네이밍은 Quantity 를 check 한다고 하는데 반환 타입은 Int 를 내뱉어 주고 있네요.
해당 함수에 맞는 반환타입인가요?

ue-mu replied:
반환 타입에 맞게 함수명을 수정하였습니다.

juun97 reviewed (commented):
잘 수정해주셨습니다 👍🏻
이로서 case 가 추가되더라도 로직은 변경되지 않고 그대로 사용할 수 있겠네요

코멘트 드린 부분만 수정되면 머지하도록 하겠습니다

Developer-Groo committed:
refactor: reduceFruitQuantity함수의 클로저 파라미터명 수정

ue-mu committed:
LocalizedError를 이용한 catch한 error 처리 및 함수 네이밍 수정

Developer-Groo committed:
refactor: 재고감소 유효성 검사 함수 네이밍 수정, 반환타입수정 및 그에 따른 reduceFruitQuantity함수 수정
